stacker/util.py

```diff
@@ -554,24 +631,108 @@ def __init__(self, sources, stacker_cache_dir=None):
+        """Make a remote S3 archive available for local use.
+
+        Args:
+            config (dict): git config dictionary
+
+        """
+        extractor_map = {'.tar.gz': TarGzipExtractor,
+                         '.tar': TarExtractor,
+                         '.zip': ZipExtractor}
+        extractor = None
+        for suffix, klass in extractor_map.iteritems():
+            if config['key'].endswith(suffix):
+                extractor = klass()
+                dir_name = self.sanitize_uri_path(
+                    "s3-%s-%s" % (config['bucket'],
+                                  config['key'][:-len(suffix)])
+                )
+                break
+
+        if extractor is None:
+            raise ValueError(
+                "Archive type could not be determined for S3 object \"%s\" "
+                "in bucket %s." % (config['key'], config['bucket'])
+            )
+
+        session = get_session(region=None)
+        extra_s3_args = {}
+        if config.get('requester_pays', False):
+            extra_s3_args['RequestPayer'] = 'requester'
+
+        # We can skip downloading the archive if it's already been cached
+        if config.get('use_latest', True):
+            try:
+                # LastModified should always be returned in UTC, but it doesn't
+                # hurt to explicitly convert it to UTC again just in case
+                modified_date = session.client('s3').head_object(
+                    Bucket=config['bucket'],
+                    Key=config['key'],
+                    **extra_s3_args
+                )['LastModified'].astimezone(dateutil.tz.tzutc())
+            except botocore.exceptions.ClientError as client_error:
+                logger.error("Error checking modified date of "
+                             "s3://%s/%s : %s",
+                             config['bucket'],
+                             config['key'],
+                             client_error)
+                sys.exit(1)
+            dir_name += "-%s" % modified_date.strftime(self.ISO8601_FORMAT)
+        cached_dir_path = os.path.join(self.package_cache_dir, dir_name)
+        if not os.path.isdir(cached_dir_path):
+            logger.debug("Remote package s3://%s/%s does not appear to have "
+                         "been previously downloaded - starting download and "
+                         "extraction to %s",
+                         config['bucket'],
+                         config['key'],
+                         cached_dir_path)
+            tmp_dir = tempfile.mkdtemp(prefix='stacker')
+            tmp_package_path = os.path.join(tmp_dir, dir_name)
+            try:
+                extractor.set_archive(os.path.join(tmp_dir, dir_name))
+                session.resource('s3').Bucket(config['bucket']).download_file(
+                    config['key'],
+                    extractor.archive,
+                    ExtraArgs=extra_s3_args
+                )
+                extractor.extract(tmp_package_path)
```

Comment:
Can you throw some debug statements around these two steps?  I feel like things could break here, and knowing where it breaks because of the debug statements will be helpful.